_specifications/specification-3-17.md

```diff
@@ -8173,6 +8188,172 @@ export interface Moniker {
 }
 ```
 
+#### <a href="#textDocument_prepareTypeHierarchy" name="textDocument_prepareTypeHierarchy" class="anchor">Prepare Type Hierarchy Request (:leftwards_arrow_with_hook:)</a>
+
+> *Since version 3.17.0*
+
+The type hierarchy request is sent from the client to the server to return a type hierarchy for the language element of given text document positions. Will return `null` if the server couldn't infer a valid type from the position. The type hierarchy requests are executed in two steps:
+
+  1. first a type hierarchy item is prepared for the given text document position.
+  1. for a type hierarchy item the supertype or subtype type hierarchy items are resolved.
+
+_Client Capability_:
+
+* property name (optional): `textDocument.typeHierarchy`
+* property type: `TypeHierarchyClientCapabilities` defined as follows:
+
+```typescript
+interface TypeHierarchyClientCapabilities {
+	/**
+	 * Whether implementation supports dynamic registration. If this is set to
+	 * `true` the client supports the new `(TextDocumentRegistrationOptions &
+	 * StaticRegistrationOptions)` return value for the corresponding server
+	 * capability as well.
+	 */
+	dynamicRegistration?: boolean;
+}
+```
+
+_Server Capability_:
+
+* property name (optional): `typeHierarchyProvider`
+* property type: `TypeHierarchyOptions | TypeHierarchyRegistrationOptions` where `TypeHierarchyOptions` is defined as follows:
+
+```typescript
+export interface TypeHierarchyOptions extends WorkDoneProgressOptions {
+	/** 
+	 * The server supports for providing an inheritance tree.
+	 */
+	inheritanceTreeSuppport?: boolean;
```

Comment:
typo in support

Reply:
I am not convinced that we should add `inheritanceTreeSuppport`  in the first version since I don't see clients to support this rendering. IMO it makes things simply harder to understand. Or do I miss something?

Reply:
My apologize for not mentioning the client behaviors about `inheritanceTreeSupport` in the protocol. I'd like to describe it in detail here:
 
When using `inheritanceTreeSupport`, the client has the same mechanism of handling response data as the other two requests. The only difference is how to set `classOnly` in `TypeHierarchySupertypesParams`. 
- If the server doesn't set `inheritanceTreeSupport` or it's `false`, the `classOnly` will always be `undefined` since the server doesn't support inheritance tree. Besides, **Inheritance Tree** button will hide in the client.
- When the server set `inheritanceTreeSupport` to `true`, the client could show **Inheritance Tree** button.
  - If the active view is inheritance tree view, after prepare request, to get the most ancestor class of the given type, the client will send several `typeHierarchy/supertypes` requests with `classOnly` in their params set to `true`. The responses include all classes in the inheritance tree of the given item, so that the client can render this tree reversally. The remaining requests work normally.
  - If the active view is not inheritance tree view, `classOnly` will always be `false` and everything works normally.
- In the server, if there is a `typeHierarchy/supertypes` request with its `classOnly` is `true`, the response should contain class only, otherwise it contains all types.

If there is no `inheritanceTreeSupport` in the protocol, since we'd like to support inheritance tree view for it's the most important type hierarchy view for Java users (might also for single inheritance language users), we have to independently implement the view in java extension instead. We might make it as a command and use `workspace.executeCommand` to trigger this view. In server, we'll have the similar logics to handle the supertype requests which return only class. Another concern is that if we can put the **Inheritance Tree** button properly. See the following screenshot:
![type2](https://user-images.githubusercontent.com/45906942/118770745-466b6700-b8b4-11eb-8e20-fcad056611d6.jpg)
Currently we could put them together since we contribute `ms-vscode.references-view`, if the right two buttons are supported in LSP and implemented in client already, can we still find a way to put them together?

What do you think of this?

Reply:
I was incorrect about what I said. I see the rendering but I am still not convinced for the modes in the API. Explaining this is simply too complicated. Why can a client not render that tree using the existing API by filtering interfaces. And what do we expect that happens if a client calls the API with `classOnly` for a item that represents a interface.

Reply:
> Why can a client not render that tree using the existing API by filtering interfaces. 

A client can render that tree by filtering interfaces indeed. It's a good tradeoff to remove `classOnly` and put all the logics in client implementation to keep the protocol clear. If we remove `classOnly` and keep `inheritanceTreeSupport` in the first version, should we define the behaviors of client? If we remove `inheritanceTreeSupport` from the first version, my concern is whether we can still keep the **inheritance tree view** (from java extension) and the other two views button (from VSCode) together. 
> And what do we expect that happens if a client calls the API with `classOnly` for a item that represents a interface.

Currently we expect calling the API with `classOnly` for an interface will return an empty array since never a class can be the supertype of an interface.

Reply:
One thing that may be worth keeping in mind is that some languages like C++ have multiple class inheritance, so `classOnly` may not imply only one supertype.

Reply:
I am not understanding why `inheritanceTreeSupport` can influence how a client can renderer a class hierarchy. Isn't this something that is under the sole control of the client and a server should never influence this?

Another problem I see with `classOnly` is that it is specific for a certain use case. It might cause problems with other languages. I could for example imagine a special view for language that have multiple class inheritance. Another view could be special for interface to see the implemented protocols. 

If we think that that performance and payload size is a problem then we should add a filter property to the params which would allows to filter hierarchy items on the server side. IMO `classOnly` is simply to specific for a specific use case.

Reply:
#### About inheritanceTreeSupport
After reading through above comments, I think current naming is causing confusion. Let me help to clarify according to my understanding: 

`ServerCapabilities.inheritanceTreeSupport` proposed by @CsCherrYY indicates whether _"the language **only** supports single inheritance through **class**"_. E.g. it's true for Java because multiple inheritance is supported through interfaces. And it's false for C++.  (@CsCherrYY correct me if I'm wrong)

"how a client can renderer a class hierarchy" mentioned by @dbaeumer  should be related to a **Client** capability if needed. For example `inheritanceTreeSupport`, `inheritanceGraphSupport` (for multiple-inheritance languages?)  etc. (@dbaeumer correct me if I'm wrong)



#### About classOnly
> If we think that that performance and payload size is a problem then we should add a filter property to the params

Agree that `classOnly` is merely a specific filter property, is there any other filtering requirement (e.g. interfaceOnly), how would a generic filter property look like?
I'm testing this proposal on Java language server, so far I don't see any performance/payload size problems as we expand the hierarchy level by level, and I'm fine to filter items on either server/client side. 

Here I only have a stupid question as I know little about client implementation details: No matter what the filtering property eventually is, if I want to support showing class view (e.g. using tree items in vscode's reference view) for Java, how does client know when to filter the items or not? Does it provide multiple entries (e.g."show supertypes", "show subtypes", "show class view", "show interface only") and send requests with different params corresponding to the entries?

Reply:
Filter properties would be properties that are available on a TypeHierarchyItem. I would say typically the  kind property.

Regarding the UI: it could either be a different command or some sort of filter buttons in the UI.

Reply:
If it's possible to keep the Class Hierarchy view available for the Java users, I'm very glad to remove `inheritanceTreeSupport` in the first version since it's indeed a little complicated. Here is another approach to support Class Hierarchy: The Java extension can contribute another button in reference-view and use a separated command to show class hierarchy.  But there is still a concern about the UI implementation.

We want to make the separated class hierarchy button align with the two existing type hierarchy buttons (supertype hierarchy and subtype hierarchy). In the current reference-view implementation, the call hierarchy uses [context value](https://github.com/microsoft/vscode-references-view/blob/870c7f3098f10090581215fbc7fcf1af23e19b40/package.json#L199) to control whether to show the button. Since the call hierarchy has only two views, it works well. But for type hierarchy, can we make the context value extendable to support possible other views? Then we can use other conditions like `enablement` to show if this button is disabled, like the current [Java implementation](https://user-images.githubusercontent.com/45906942/118770745-466b6700-b8b4-11eb-8e20-fcad056611d6.jpg).

@dbaeumer Could you give any suggestions?

Reply:
Update: Another concern is about the default view. Currently the default type hierarchy view for java users is Class Hierarchy view, which is most frequently used by java developers. However, the implementation of type hierarchy in [vscode-references-view](https://github.com/microsoft/vscode-references-view) would be based on the protocol, supporting subtype hierarchy view and supertype hierarchy view only. Could it be possible to change the default view of type hierarchy in another extension? e.g. Java extension.